www/js/survey/enketo-survey-service.js

```diff
@@ -0,0 +1,213 @@
+   *  loaded: {
+   *    form: string;
+   *    model: string;
+   *  };
+   *  opts: {
+   *    trip: object;
+   *    uuid: string;
+   *  };
+   * }}
+   */
+
+  const DATA_KEY = 'manual/survey_response';
+  const ENKETO_SURVEY_CONFIG_PATH = 'json/enketoSurveyConfig.json';
+  /** @type {EnketoSurveyState} _state */
+  let _state;
+
+  /**
+   * _lazyLoadConfig load enketo survey config. If already loaded, return the cached config
+   * @returns {Promise<EnketoSurveyConfig>} enketo survey config
+   */
+  function _lazyLoadConfig() {
+    if (_state.config !== null) {
+      return Promise.resolve(_state.config);
+    }
+    return $http.get(ENKETO_SURVEY_CONFIG_PATH).then(configRes => {
+      _state.config = configRes.data;
+      return _state.config;
+    });
+  }
+
+  /**
+   * _loadForm load enketo form
+   * @param {{
+   *  instanceStr?: string;
+   *  submitted?: boolean;
+   *  external?: string[];
+   *  session?: object;
+   * }} opts enketo form options
+   * @returns {string[]} load errors
+   */
+  function _loadForm(opts = {}) {
+    const formSelector = 'form.or:eq(0)';
+    const data = {
+      // required string of the default instance defined in the XForm
+      modelStr: _state.loaded.model,
+      // optional string of an existing instance to be edited
+      instanceStr: opts.instanceStr || null,
+      submitted: opts.submitted || false,
+      external: opts.external || [],
+      session: opts.session || {}
+    };
+    _state.form = new $window.FormModule(formSelector, data, {});
+    return _state.form.init();
+  }
+
+  /**
+   * _resetState reset state object to its initial value
+   */
+  function _resetState() {
+    _state = {
+      config: null,
+      name: null,
+      form: null,
+      formLocation: null,
+      loaded: {
+        form: null,
+        model: null,
+      },
+      opts: {
+        trip: null,
+        uuid: null,
+      },
+    };
+  }
+
+  /**
+   * _restoreAnswer restore the most recent answer for the survey
+   * @param {EnketoAnswer[]} answers survey answers
+   * @returns {string} answer string
+   */
+  function _restoreAnswer(answers) {
+    let answer = null;
+    answers = EnketoSurveyAnswer.filterByNameAndVersion(_state.name, answers);
+    if (!answers.length) {
+      return null;
+    }
+    if (_state.opts.trip) {
+      answer = DiaryHelper.getUserInputForTrip(_state.opts.trip, undefined, answers);
+      if (answer) {
+        return answer.data.xmlResponse;
+      }
+    }
+    if (_state.opts.uuid) {
+      answers = answers.filter(answer => answer.data.user_uuid === _state.opts.uuid);
+      answer = answers.length ? answers[0] : null;
+    }
+    return answer ? answer.data.xmlResponse : null;
+  }
+
+  /**
+   * _saveData save survey answer data
+   * @returns {Promise<EnketoAnswerData>} answer data
+   */
+  function _saveData() {
+    /** @type {DOMParser} xmlParser */
+    const xmlParser = new $window.DOMParser();
+    const xmlResponse = _state.form.getDataStr();
```

Comment:
nit: Do you plan to store the JSON response as well? Of course, totally fine to do that as a separate enhancement.

Reply:
Having it in JSON will help with the deployer dashboard since it can be automagically displayed just like the various `confirm` fields in the trip table.

Reply:
We need to store it as  XML doc to make it compatible with Enketo library when we restore the answer. But storing JSON along with XML would be nice as you suggested. I will to this one later.